Apps/Src/ReadCarCAN.c

```diff
@@ -70,99 +64,41 @@ static void updateSaturation(int8_t chargeMessage){
 }
 
-// exception struct callback for charging disable, kills contactors and turns of display
-static void callback_disableContactors(void){
-    // Kill contactors 
+// exception struct callback for charging disable, kills contactor and turns off display
+static void callback_chargingDisable(void){ 
+    // mark regen as disabled and update saturation
+    updateSaturation(-1);
+
+    // Kill contactor
     Contactors_Set(ARRAY_CONTACTOR, OFF, true);
-    Contactors_Set(ARRAY_PRECHARGE, OFF, true);
 
     // Turn off the array contactor display light
     UpdateDisplay_SetArray(false);
 }
 
-// helper function to disable charging
-// Turns off contactors by signaling fault state
-static inline void chargingDisable(void) {
-    // mark regen as disabled
-    chargeEnable = false;
-
-    // kill contactors 
-    exception_t readBPSError = {.prio=2, .message="read BPS error", .callback=callback_disableContactors};
-    assertExceptionError(readBPSError);
-}
-
-// helper function to call if charging should be enabled
-static inline void chargingEnable(void) {
-    OS_ERR err;
-    CPU_TS ts;
-
-    // mark regen as enabled
-    chargeEnable = true;
-
-    // check if we need to run the precharge sequence to turn on the array
-    bool shouldRestartArray = false;
-
-    OSMutexPend(&arrayRestartMutex, 0, OS_OPT_PEND_BLOCKING, &ts, &err);
-    assertOSError(OS_READ_CAN_LOC,err);
-
-    // if the array is off and we're not already turning it on, start turning it on
-    shouldRestartArray = !restartingArray && (Contactors_Get(ARRAY_CONTACTOR)==OFF);
-
-    // wait for precharge for array 
-    if(shouldRestartArray){
-
-            restartingArray = true;
-
-            // Wait to make sure precharge is finished and then restart array
-            OSTmrStart(&prechargeDlyTimer, &err);
-            assertOSError(OS_READ_CAN_LOC, err);
-            
-        }
-    
-
-    OSMutexPost(&arrayRestartMutex, OS_OPT_NONE, &err);
-    assertOSError(OS_READ_CAN_LOC,err);
+// Nested function because same function needs to be executed
+// however the function declarations require different arguments
+static void chargingDisable(void *p_tmr, void *p_arg){
+    callback_chargingDisable();
 }
 
 /**
  * @brief Callback function for the precharge delay timer. Waits for precharge and then restarts the array.
- * @param p_tmr pointer to the timer that calls this function, passed by timer
- * @param p_arg pointer to the argument passed by timer
- * 
 */
+Minion_Error_t Merr;
 static void arrayRestart(void *p_tmr, void *p_arg){
-    Minion_Error_t Merr;
     if(chargeEnable){    // If regen has been disabled during precharge, we don't want to turn on the main contactor immediately after
-        Contactors_Set(ARRAY_CONTACTOR, (Minion_Read_Pin(IGN_1, &Merr)), false); //turn on array contactor if the ign switch lets us
+        Contactors_Set(ARRAY_CONTACTOR, (Minion_Read_Pin(IGN_1, &Merr)), false); // Turn on array contactor if the ign switch lets us
```

Comment:
I would really like to get rid of the Minion_Read_Pin here, that was the main idea behind this issue ticket in the first place. Can we use a static variable and only read the pin in the main loop once, so that we avoid certain edge cases?